big name changes flg_rba and flb_abr

hmb1 opened the pull request:
Have used a more consistent nameing convention.
flg_rba if set now calculates the ratios before average
flg_abr if set now calculates the averages before ratio (formula) step

attached is Netcdf meta-data from a typical run on ROGER

@czender and @remotesensinglab please comment and review
Many Thanks
..Henry

[indices.txt](https://github.com/terraref/extractors-hyperspectral/files/843973/indices.txt)

dlebauer commented:
Here is the metadata formatted for readability https://gist.github.com/dlebauer/70af4579ab62dbd58388bfcd6069e1c3

A few questions:

* What are variables on lines 7-37 are (flg_* and iR* ...) 
  * should they be defined or removed?
* doesn't the source have variables float `latitude (y, x)` and float `longitude(y, x)`? Can these be retained?
* why are all of the standard names `surface_albedo`?
   * for wavelengths, these don't make sense without information about the wavelength 
   * for indices, the values aren't actually surface_albedo
       * standard_name field should match the standard name fields in BETYdb, e.g. [NDVI](https://terraref.ncsa.illinois.edu/bety/api/beta/variables?key=9999999999999999999999999999999999999999&limit=none&name=NDVI)
        * please make sure that each of the variable names link to a variable in BETYdb with the API call `https://terraref.ncsa.illinois.edu/bety/api/beta/variables?key=9999999999999999999999999999999999999999&limit=none&name=$VARIABLE_NAME`
   * is there a reason to keep the wavelength layers? Aren't these just copied from the calibrated (Level 1) data product?

dlebauer commented:
Following from above, the label, long_name, and standard_name should be unique for each iindex. And should map to BETYdb fields thus

long_name --> BETYdb label 
standard_name --> standard_name

That said, I now see that I need to clean up the BETYdb variables table. I will go ahead and do that. Could these values be queried dynamically?

gsrohde commented:
@dlebauer If you could give me a heads up on what sort of cleanup you are doing to the variables table, that would be helpful—most especially if you are planning to make any changes to `name` column values, since this will affect pending data uploads.

dlebauer commented:
@gsrohde sorry ... I was about to paste queries for you but got carried away. The names haven't changed

@hmb1 for metadata, I would suggest that this"

```yml
	float TCARI ;
		TCARI:_FillValue = 1.e+36f ;
		TCARI:description = "Transformed chlorophyll absorption in reflectance index | 3.0f*(  (R700-R670)-0.2f * (R700-R550) * (R700/R670)  ) | Haboudane et al. (2002)" ;
		TCARI:label = "Reflectance Index" ;
		TCARI:long_name = "Reflectance of image" ;
		TCARI:max = 16.27735f ;
		TCARI:min = -571.2423f ;
		TCARI:notes = "Transformed chlorophyll absorption in reflectance index | TCARI" ;
		TCARI:standard_name = "surface_albedo" ;
		TCARI:type = "trait" ;
		TCARI:units = "1" ;
```

Could be changed to 

(reducing the number of fields and updating the contents following https://github.com/terraref/reference-data/issues/105

```yml
	float TCARI ;
		TCARI:_FillValue = 1.e+36f ;
		TCARI:description = "Transformed chlorophyll absorption in reflectance index | 3.0f*(  (R700-R670)-0.2f * (R700-R550) * (R700/R670)  ) | Haboudane et al. (2002)" ;
		TCARI:long_name = "Transformed Chlorophyll Absorption Ratio TCARI" ;
		TCARI:max = 16.27735f ;
		TCARI:min = -571.2423f ;
		TCARI:standard_name = "transformed_chlorophyll_absorption_in_reflectance_index" ;
		TCARI:units = "1" ;
```

hmb1 commented:
Hi All,
 here is the latest meta-data - in the first version some of the meta-data was being overwritten.
[Uploading new_ind.txt…]()
 
1) Charlie have removed the extraneous global attributes
2) David , the extraneous wavelengths 'R???' and indexes 'iR???'  - I will remove by adding another ncks extraction step to the workflow -(once we have nailed down the meta-data)

3)  I have added the coordinate variables x,y,wavelength to the output is this this sufficient ? 
     In the input file are vars like 'lat_img_nw',  'x_img_se', y_img_se' ...   do you want these as well ?

4)  Each var name now has the requested attribute fields ,
     long_name,standard_name, description, notes, units, label  
    There are typographic differences between these fields and the ones in BettyDB ?

     I could write a python script that queries BettyDB periodically  for each index and writes an ncap2 script, that could be run after the main nco script.  Would this work for you ?

....Henry

hmb1 commented:
Hi All,
 here's the link again to the cdl file. 
 
[new_ind.txt](https://github.com/terraref/extractors-hyperspectral/files/855018/new_ind.txt)

 
 ...Henry

hmb1 commented:
@dlebauer 
Hi David,
 have now got a python script for extracting data from the json Url mentioned above.
 Is there such a query that will give me all the index names - then  I can complete the script 
Thanks

czender commented:
Each iteration looks better. There are still issues that keep this from being finished. I will let @dlebauer decide how frequently to merge these iterative improvements, though I think we should do it at least once a week.
1. wvl_clb: this variable does not belong in this file.
2. I don't know what "notes" and "label" are for, I assume these values come from someone/somewhere trustworthy.  However, I do know that "long_name" should be a short English description of the variable suitable for use as an axis label. Hence it must differentiate between variables yet there are many instances of "long_name" that have the same value, "Reflectance of image". It seems like this was just copied, and does not convey the true meaning of the field. Please correct if possible. Note how long_name in foo_pxl is often incorrect but long_name in foo is correct. And mostly foo_pxl and foo should have identical attributes (except for units, min, and max). I would have thought that with attribute propagation the attributes would be identical unless explicitly overwritten. If you are uncertain what to use for "long_name", please replace it with "fxm" and I or David will fill-it-in.
3. CF-compliance: Please add "cell_methods" to the foo (averaged) variables. If ncwa did the averaging this would be done automatically, and result in something like "cell_methods = "x, y: mean"
4. @dlebauer will chime-in on how much coordinate information to retain.
5.  If @delbauer does not the 'R???' and index 'iR???' variables, then can we not remove them by declaring them as RAM variables in the script rather than running another command?

dlebauer commented:
> I could write a python script that queries BettyDB periodically for each index and writes an ncap2 script, that could be run after the main nco script. Would this work for you ?

I think these should be invariant with time, so we should fix them in BETYdb. If anything send an error if an inconsistency is found 

CZ for 4) I hand off to @max-zilla  5) yes that seems sensible

hmb1 commented:
Hi @ac-gl-cris
 I have added your indices and grouped them as per your comments 
 Some of the groups are switched off by default as they use SWIR wavelengths  and possibly VNIR wavelengths -later we can use them.

// March 2017 - this contains SWIR wavelengths 
  *flg_cni=0s; // [flg] // Canopy Nitrogen Indices

// March 2017 - this contains SWIR wavelengths - so default is OFF
  *flg_sci=0s; // [flg] // Dry or Senescent Carbon Indices 

// March 2017 - this contains SWIR wavelengths -so default is OFF
  *flg_cwci=0s; // [flg] // Canopy Water Content Indices

// March 2017 - this contains SWIR wavelengths -so default is OFF
  *flg_dwsi=0s; // [flg] // Spectral Disease Indices

A couple of your indices use `pow()` - In eta the pow()  arg had only one argument so I assumed us wanted to square and added '2'. if you could check the other usages of pow() that would be grand.

...Henry

hmb1 commented:
Hi all,
 The new indices only have long_name   
 1) add the new indices to BETYDB. And have  script that extracts them to an cnap2 script
 2)  add then to the main ncap2 script

 Please review the netCDF meta-data: 
[n2_ind.txt](https://github.com/terraref/extractors-hyperspectral/files/879085/n2_ind.txt)

...Henry

hmb1 commented:
cz 4)  If David does not want the var_pxl data then i suggest an additional ncks step with removes the var_pxl variables and the R??? , iR??? data
         if David wishes to retain the var_pxl data then I will recode R??? and iR??? as ram variables as you suggest
...Henry

ac-gl-chris commented:
Hi @Henry
ah sorry, all usages of pow should be power of 2, so in eta pow(x,2) is correct. The rest looks correct as well.
Chris

hmb1 commented:
Hi @ac-gl-chris ,
 whats your view on the meta-data ? long_name/ standard_name/label/ etc 
Is that iyour domain ?
...HEnry

ac-gl-chris commented:
Appears a bit redundant but I see the idea.
However, "notes" seems to be "long_name | abbreviation" and "description" is "long_name | formular | reference to a paper". The formular could get a bit complex. I would go with "long_name" as is, "notes" as abbreviation and "description" as reference but I don't know how the meta data is presented to the user. Maybe it makes perfect sense =)
Should I apply it to the new indices?

Chris

hmb1 commented:
HI @ac-gl-chris,
  better check with @dlebauer first -
...Henry

czender commented:
@hmb1 your response above about var_pxl data seems like a fine way to proceed. We need @dlebauer to clarify whether the var_pxl data (i.e., the full images of each statisitic at every pixel, before spatial averaging) will ever be desired in the L2 output. Perhaps it will be desired in L2 but not in L3, in which case the elimination step would take place at a later stage.

dlebauer commented:
@remotesensinglab can you make the call?

remotesensinglab commented:
Yes, I think it is desired at L2 but not at L3. We need pixel level images and data at L2. Thanks

dlebauer commented:
@hmb1 just to clarify ... L1 is full resolution of the individual bands and L2 is full resolution on indices.

dlebauer commented:
OK I put the latest version of the metadata here: https://gist.github.com/dlebauer/70af4579ab62dbd58388bfcd6069e1c3#file-ncdump2-yml

It does not look like you have addressed comments from @czender in https://github.com/terraref/extractors-hyperspectral/pull/20#issuecomment-287888563, in particular

* wvl_clb: this variable does not belong in this file.
* we don't need to keep the individual wavelengths, just the indices
  * I gather these are the R and iR variables? what does 'i' mean? Can these all be dropped?
* what are the foo_pxl values for? Can these be dropped?
* Please add "cell_methods" to the foo (averaged) variables. (that is from @czender I don't know what it means but I assume it provides information about the algorithm used?)

czender commented:
And please select long_name's that make a user's life easy, e.g., R123* field long_names should include the wavelength, i.e., "Reflectance of image at 540 nm" to differentiate them from one another, and so a user making an automatic plot does not have to manually add the wavelength to the title.  In other words, go through the variables and metadata and ask whether each is necessary and sufficient.

hmb1 commented:
HI Charlie,
 have used ram vars for  reflectances --  need to watch this as am currently extracting 49 out of 955 
 Anyway this is the meta-data: 
[new_ind2.txt](https://github.com/terraref/extractors-hyperspectral/files/893183/new_ind2.txt)

Can further simply script  by using atoi() to get wavelengths from name e.g  
@nm="R452" 
@val=atoi(@nm(1:))*1.e-10d;  // @val=4.52e-7

atoi()/atol()  not working in "nco/4.6.5-alpha02"  but is now working in latest repo

...Henry

hmb1 commented:
have added "cell_methods"   to  fmc_all_cls.cc -agg_cls() - 
The default for cell_methods is on
...Henry

czender commented:
These look good to me but I did not notice that cell_method metadata was present. Is this because the NCO on roger needs to be updated to latest snapshot?

hmb1 commented:
HI Charlie,
 cell_methods() is still in bleeding edge repo as is atoi() and atol()
 In the latest version version of  "hyperspectral_indices_make.nco" I changes the "pow(blah-blah,2)" to sqr(blah-blah). if everything looks OK pls commit to master so that Chris can add the meta -data
...Henry

czender commented:
If you're happy with the atoi(), atof(), cell_methods in the current repo, I'll cut a new tag that Yan Liu can install on Roger so the automatically generated metadata will start to appear.

hmb1 commented:
OK sure go for it ...Henry


On 05/04/17 14:46, Charlie Zender wrote:
>
> If you're happy with the atoi(), atof(), cell_methods in the current 
> repo, I'll cut a new tag that Yan Liu can install on Roger so the 
> automatically generated metadata will start to appear.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/terraref/extractors-hyperspectral/pull/20#issuecomment-291866302>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGIX8CUX4a-cZ12RnL8luuKel4lqxi2Xks5rs5tNgaJpZM4MdsYd>.
>

czender commented:
OK this is ready. @yanliu-chn please install NCO 4.6.6-alpha03 on roger as the default module. It contains a new feature that should automatically generate cell_methods metadata useful in the hyperspectral L2 and L3 products. @hmb1 once this is done please post an updated view of the metadata from the L2 indices file, that will presumably show cell_methods is present in the averaged variables.

yanliu-chn commented:
done. NCO 4.6.6-alpha03 on ROGER is installed. the default module load changed to this version. to use:

`module purge`
`module load antlr-2.7.7 udunits gdal-stack-2.7.10 nco`

hmb1 commented:
latest meta-data
nb there is a minor issue that needs fixing -will do on ROGER ASAP
[new_ind.txt](https://github.com/terraref/extractors-hyperspectral/files/899928/new_ind.txt)
...HEnry

czender commented:
There should be no cell_methods for the pxl variables because they have not been averaged, right?
yet they say 

cell_methods = "y, x: maximum"

is this a flaw in the ncap2 implementation? Please re-read the CF document and fix implementation.

hmb1 commented:
It a problem with attribute propagation - its odd that it hasn't 
happened before -
If got an ugly fix so i shall sleep on it

On 05/04/17 20:44, Charlie Zender wrote:
>
> There should be no cell_methods for the pxl variables because they 
> have not been averaged, right?
> yet they say
>
> cell_methods = "y, x: maximum"
>
> is this a flaw in the ncap2 implementation? Please re-read the CF 
> document and fix implementation.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/terraref/extractors-hyperspectral/pull/20#issuecomment-291975058>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGIX8K-UJLHaIe4yfKZmYW_eGqjq4ALgks5rs-8NgaJpZM4MdsYd>.
>
